File: packages/axon-core-py/src/axon_core/production.py
from __future__ import annotations
from typing import Literal, Optional
from pydantic import BaseModel, Field

ProdStatus = Literal[
    "created", "estimated", "scheduled", "released", "started",
    "reportedFinished", "ended", "cancelled",
]


class OperationStep(BaseModel):
    oprNum: int
    oprId: str
    oprPriority: int
    jobIdentification: Optional[str] = None
    capacityCheck: Optional[bool] = None
    resourceId: Optional[str] = None
    oprDateTime: Optional[str] = None
    setupTimeSec: Optional[float] = None
    runTimePerPieceSec: Optional[float] = None


class ProductionOrder(BaseModel):
    id: str
    prodId: str
    itemId: str
    qtyPlanned: float
    qtyReported: float = 0.0
    status: ProdStatus
    operations: list[OperationStep] = Field(default_factory=list)
    batchNumber: Optional[str] = None
    inventBatchId: Optional[str] = None
    createdAt: Optional[str] = None
    scheduledStart: Optional[str] = None
    scheduledEnd: Optional[str] = None


class KanbanCard(BaseModel):
    id: str
    itemId: str
    qty: float
    status: Literal["empty", "in-transit", "filled", "consumed"]
    fromLocation: str
    toLocation: str
    productionOrderId: Optional[str] = None

File: packages/axon-core-ts/src/production.ts
/**
 * Production order + operation step model. Field names mirror Microsoft D365
 * F&O Production module so that the ShopFloor API integration can
 * round-trip without translation. These are public Microsoft schema names.
 */
export type ProdStatus =
  | "created"
  | "estimated"
  | "scheduled"
  | "released"
  | "started"
  | "reportedFinished"
  | "ended"
  | "cancelled";

export interface OperationStep {
  oprNum: number;
  oprId: string;
  oprPriority: number;
  jobIdentification?: string;
  capacityCheck?: boolean;
  resourceId?: string;
  oprDateTime?: string; // ISO-8601
  setupTimeSec?: number;
  runTimePerPieceSec?: number;
}

export interface ProductionOrder {
  id: string;
  prodId: string;
  itemId: string;
  qtyPlanned: number;
  qtyReported: number;
  status: ProdStatus;
  operations: OperationStep[];
  batchNumber?: string;
  inventBatchId?: string;
  createdAt?: string;
  scheduledStart?: string;
  scheduledEnd?: string;
}

export interface KanbanCard {
  id: string;
  itemId: string;
  qty: number;
  status: "empty" | "in-transit" | "filled" | "consumed";
  fromLocation: string;
  toLocation: string;
  productionOrderId?: string;
}
